src/operation/valid/MakeValid.cpp

```diff
@@ -0,0 +1,344 @@
+/**********************************************************************
+ *
+ * GEOS - Geometry Engine Open Source
+ * http://geos.osgeo.org
+ *
+ * Copyright 2009-2010 Sandro Santilli <strk@kbt.io>
+ * Copyright (C) 2019 Even Rouault <even.rouault@spatialys.com>
+ *
+ * This is free software; you can redistribute and/or modify it under
+ * the terms of the GNU Lesser General Public Licence as published
+ * by the Free Software Foundation.
+ * See the COPYING file for more information.
+ **********************************************************************
+ *
+ * Ported from rtgeom_geos.c from
+ *   rttopo - topology library
+ *   http://git.osgeo.org/gitea/rttopo/librttopo
+ * with relicensing from GPL to LGPL with Copyright holder permission.
+ *
+ **********************************************************************/
+
+#include <geos/operation/valid/MakeValid.h>
+#include <geos/operation/valid/IsValidOp.h>
+#include <geos/operation/polygonize/BuildArea.h>
+#include <geos/geom/Geometry.h>
+#include <geos/geom/GeometryCollection.h>
+#include <geos/geom/GeometryFactory.h>
+#include <geos/geom/LineString.h>
+#include <geos/geom/Point.h>
+#include <geos/geom/Polygon.h>
+#include <geos/geom/MultiLineString.h>
+#include <geos/geom/MultiPolygon.h>
+#include <geos/util/UniqueCoordinateArrayFilter.h>
+#include <geos/util/UnsupportedOperationException.h>
+
+// std
+#include <cassert>
+#include <algorithm>
+#include <utility>
+#include <vector>
+
+#ifdef _MSC_VER
+#pragma warning(disable:4355)
+#endif
+
+using namespace geos::geom;
+
+namespace geos {
+namespace operation { // geos.operation
+namespace valid { // geos.operation.valid
+
+/*
+ * Fully node given linework
+ */
+static std::unique_ptr<geom::Geometry> 
+GEOS_nodeLines(const geom::Geometry* geom)
```

Comment:
How about something like `nodeLineWithFirstCoordinate` ?

Reply:
This should be just GEOSNode call. When I tried replacing it with GEOSNode in PostGIS it crashed painfully. Inside GEOS I expect it to be replaced with GEOSNode and GEOSNode fixed to not crash.

Reply:
GEOSNode is in the C API; it's not callable here.

Reply:
its C++ internals should be called then? not a new implementation. It's just a side effect that Union does the noding.

Reply:
Sure, there may be an other way to implement it. But I see no reason that Even should fix unticketed bugs as part of this PR.

Reply:
Renamed as nodeLineWithFirstCoordinate() as suggested.

GEOSNode() just calls geos::noding::GeometryNoder::node(). I tried this, but unfortunatelly this doesn't give the same results as the current implementation. For example it doesn't return POINT(0 0) for LINESTRING(0 0,0 0), but MULTILINESTRING((0 0,0 0))